rotkehlchen/chain/ethereum/modules/uniswap/utils.py

```diff
@@ -253,3 +273,319 @@ def historical_uniswap_v2_lp_price(
+            )
+        except RemoteError as e:
+            log.error(UNISWAP_V3_ERROR_MSG.format('nft contract positions', str(e)))
+            continue
+        positions = [
+            nft_manager_contract.decode(
+                result=data[1],
+                method_name='positions',
+                arguments=[tokens_ids[index]],
+            )
+            for index, data in enumerate(positions_multicall) if data[0] is True
+        ]
+        pool_addresses = [
+            compute_pool_address(
+                factory_address=UNISWAP_V3_FACTORY_ADDRESS,
+                token0_address=position[2],
+                token1_address=position[3],
+                fee=position[4],
+            )
+            for position in positions
+        ]
+        pool_contracts = [
+            EthereumContract(
+                address=pool_address,
+                abi=UNISWAP_V3_POOL_ABI,
+                deployed_block=UNISWAP_V3_DEPLOYED_BLOCK,
+            )
+            for pool_address in pool_addresses
+        ]
+        try:
+            slots_0_multicall = multicall_2(
+                ethereum=ethereum,
+                require_success=False,
+                calls=[
+                    (entry[0], entry[1].encode('slot0'))
+                    for entry in zip(pool_addresses, pool_contracts)
+                ],
+            )
+        except RemoteError as e:
+            log.error(UNISWAP_V3_ERROR_MSG.format('pool contract slot0', str(e)))
+            continue
+        slots_0 = [
+            entry[0].decode(entry[1][1], 'slot0')
+            for entry in zip(pool_contracts, slots_0_multicall) if entry[1][0] is True
+        ]
+        tokens_a = [
+            ethereum.get_basic_contract_info(to_checksum_address(position[2]))
+            for position in positions
+        ]
+        tokens_b = [
+            ethereum.get_basic_contract_info(to_checksum_address(position[3]))
+            for position in positions
+        ]
+        price_ranges = [
+            calculate_price_range(
+                tick_lower=entry[2][5],
+                tick_upper=entry[2][6],
+                decimal_0=entry[0]['decimals'],
+                decimal_1=entry[1]['decimals'],
+            )
+            for entry in zip(tokens_a, tokens_b, positions)
+        ]
+        amounts_0 = [
+            calculate_amount_0(
+                tick=entry[0][1],
+                tick_bottom=entry[1][5],
+                tick_top=entry[1][6],
+                liquidity=entry[1][7],
+                decimals=entry[2]['decimals'],
+            )
+            for entry in zip(slots_0, positions, tokens_a)
+        ]
+        amounts_1 = [
+            calculate_amount_1(
+                tick=entry[0][1],
+                tick_bottom=entry[1][5],
+                tick_top=entry[1][6],
+                liquidity=entry[1][7],
+                decimals=entry[2]['decimals'],
+            )
+            for entry in zip(slots_0, positions, tokens_b)
+        ]
```

Comment:
Same here. I am not going to rewrite the code for you ...

But look what you are doing. You are iterating different zips of different lists and each time making yet another iteation of everything.

Why not combine those?